File: py/src/aoc2025/day11.py
from collections.abc import Container
from typing import TypeVar


def _parse(input: str) -> dict[str, list[str]]:
    return {
        line[: (i := line.index(":"))]: line[i + 1 :].split()
        for line in input.splitlines()
        if ":" in line
    }


T = TypeVar("T")


def _paths(graph: dict[T, Container[T]], src: T, dst: T) -> int:
    cache = {dst: 1}

    def get(key: T) -> int:
        if key not in cache:
            cache[key] = sum(map(get, graph.get(key, ())))
        return cache[key]

    return get(src)


def _part1(input: str) -> int:
    return _paths(_parse(input), "you", "out")


def _part2(input: str) -> int:
    graph = _parse(input)
    return _paths(graph, "svr", "dac") * _paths(graph, "dac", "fft") * _paths(
        graph, "fft", "out"
    ) + _paths(graph, "svr", "fft") * _paths(graph, "fft", "dac") * _paths(
        graph, "dac", "out"
    )


parts = (_part1, _part2)

File: kt/src/main/kotlin/com/github/ephemient/aoc2025/Day11.kt
package com.github.ephemient.aoc2025

object Day11 {
    private fun parse(input: String): Map<String, List<String>> = input.lines().mapNotNull { line ->
        val split = line.indexOf(':')
        if (split == -1) null else line.take(split) to line.drop(split + 1).split(' ')
    }.toMap()

    private fun Map<String, Iterable<String>>.paths(src: String, dst: String): Int {
        val cache = mutableMapOf(dst to 1)
        return DeepRecursiveFunction { key ->
            cache.getOrPut(key) { get(key)?.sumOf { this.callRecursive(it) } ?: 0 }
        }(src)
    }

    fun part1(input: String): Int = parse(input).paths("you", "out")

    fun part2(input: String): Long = with(parse(input)) {
        paths("svr", "dac").toLong() * paths("dac", "fft") * paths("fft", "out") +
                paths("svr", "fft").toLong() * paths("fft", "dac") * paths("dac", "out")
    }
}
